### PR TITLE
Implement DIFF data explanation algorithm with risk ratios

### LINKED ISSUE #8: Wrapper to support grouping sts on SQLite/DuckDB/databases that don't natively support cubes and grouping sets

### LINKED ISSUE #10: Talk to Eugene Wu once candidate generation works
# Notes for conversation
* No need to read code --- I can explain it all
* So far, I've implemented two things
  * [column statistics](https://github.com/marcua/datools/blob/main/datools/table_statistics.py#L142) --- given a table, identifies range- and set-valued columns. For range-valued columns, identifies percentile bucket boundaries (e.g., 3 values representing [start_of_first, start_of_second, start_of_third] bucket values). For set-valued columns, identifies the (e.g., 100) most popular values.
  * Based on the column statistics, a [candidate predicate generator](https://github.com/marcua/datools/blob/main/datools/scorpion/explanations.py#L18) that, for each column statistic, generates predicates. For range-valued columns, this means predicates of the form `column >= start_of_first AND column < start_of_second`. For set-valued columns, this means predicates of the form `column = popular_value`.
* With those primitives, we can implement the fun part! Some thoughts
  * I liked your suggestion to implement https://homes.cs.washington.edu/~suciu/main_explanation.pdf ("the UW paper") instead of https://dspace.mit.edu/bitstream/handle/1721.1/89076/scorpion-vldb13.pdf?sequence=1&isAllowed=y ("the Scorpion paper"), because I can then push as much of the heavy lifting in to the DB without relying on external libraries for decision trees, etc. One hitch in implementing the UW paper is it relies heavily on cubes, which several DBs don't natively implement (especially SQLite and DuckDB, which I'm targeting for tests before supporting a broader set of databases). This isn't a huge problem: while I await a conversation with you, [I'll implement a wrapper](https://github.com/marcua/datools/issues/8) that implements grouping sets/cubes by way of a bunch of UNION ALL of GROUP BY combinations.
  * One thing to align on early: **what's the API**?
    * The UW paper says the user provides a list of queries with individual aggregates that can be arithmetically combined, along with a high/low direction indicator:
    image.
    * The Scorpion paper says the user provides an annotation over the aggregate query that separates the query into a hold-out set and a set of outliers that are annotates with high/low direction indicators:
    image.
  * Once we agree on the API, I also **need help deciphering the UW algorithm for a single table**. Namely
    * What's the difference between the two metrics `\mu_{aggr}` and `\mu_{interv}`?
    * Can we work through implementing the cube from the UW paper on the sensor example in the Scorpion paper since I've got that "sample dataset" implemented in [the tests](https://github.com/marcua/datools/blob/main/tests/test_scorpion.py#L17)?

### PR DESCRIPTION
**Pull request by marcua**
Resolves #10 
Resolves #8

### CODE REVIEW

**Comment by sirrice** [comment]
datools.sqlalchemy_util:39

```
rows = results.first().num_rows   #_mapping['num_rows']
```

File: datools/scorpion/explanations.py [hunk 1]
```diff
@@ -1,80 +0,0 @@
...
-from datools.models import Constant
-from datools.models import Operator
-from datools.models import Predicate
-from datools.models import Table
-from datools.table_statistics import column_statistics
-from datools.table_statistics import RangeValuedStatistics
-from datools.table_statistics import SetValuedStatistics
-
-
-def _single_column_candidate_predicates(
```
[the remark this answers is not in the document]
Comment (marcua) [reply datools/scorpion/explanations.py:18]: Nevermind---It's in algorithms.py

File: datools/explanation/algorithms.py [hunk 2]
```diff
@@ -0,0 +1,232 @@
...
+    :param group_bys: A list of columns to group on.
+    :param aggregate: An aggregate on which we identify an outlier.
+    :param outlier_predicates: Filters on the result that
+        exhibit unexpected conditions.
+    :param holdout_predicates: Filters on the result that
+        exhibit normal conditions.
+    """
+    candidates = _single_column_candidate_predicates(
+        engine, table, group_bys, aggregate)
+    return tuple(candidates)
```
Comment (marcua) [inline datools/explanation/algorithms.py:94]: Remove me

**Comment by marcua** [comment]
Thank you @sirrice! Fixed!

**Pull request merged**